### PR TITLE
Materialize expression_statement nodes

### PR DESCRIPTION
**Pull request by adonovan**
This change causes the parser to materialize expression_statement nodes for calls and receive expressions that appear as a statement, so that every child of a block is a statement (or comment).

REVIEWERS: why does test --update make a bunch of unnecessary whitespace changes to the corpus/*.txt files? I'm using ./node_modules/tree-sitter-cli/tree-sitter (v0.19.5) for this step. Should I be doing something different?

Checklist:
- [ ] All tests pass in CI.
- [ ] There are sufficient tests for the new fix/feature.
- [ ] Grammar rules have not been renamed unless absolutely necessary.
- [ ] The conflicts section hasn't grown too much.
- [ ] The parser size hasn't grown too much (check the value of STATE_COUNT in src/parser.c).

### CODE REVIEW

**Comment by maxbrunsfeld** [comment]
> every child of a block is a statement

This seems like a good change.

> why does test --update make a bunch of unnecessary whitespace changes?

It is reformatting the S-expressions. I'm ok with committing these changes. For now, we'll just need to verify that no `ERROR` or `MISSING` nodes have been introduced.

**Comment by maxbrunsfeld** [comment]
Can you upgrade to the latest Tree-sitter and regenerate with that? Currently, the latest version is 0.20.6.

File: grammar.js [hunk 1]
```diff
@@ -79,6 +79,12 @@ module.exports = grammar({
 
   word: $ => $.identifier,
 
+  // Go is essentially LL(1), but this grammar intentionally accepts
```
[the remark this answers is not in the document]
Comment (adonovan) [reply grammar.js:82]: The main exception I'm aware of is fields `struct { a, b, c int; d, e int }` and parameters `f(a, b, c int, d, e int)`, where you need to maintain an unbounded stack of a, b, c identifiers before you can reduce when you hit the int. Obviously those cases are not in LL(k) but they are easily handled in a recursive descent parser. I guess it accounts for most of the conflicts here. Do you know of others? Things may have gotten gnarlier with generics.

[the remark this answers is not in the document]
Comment (maxbrunsfeld) [reply grammar.js:82]: I don’t think the conflicts are due to allowing top-level statements. As far as I remember, most or all of them are due to ambiguities between types and expressions, due to the syntax for type conversions resembling calls.

[the remark this answers is not in the document]
Comment (aryx) [reply grammar.js:82]: The problem you mention for the fields and parameters is fine. The original yacc grammar for Go was handling that fine. 

I don't remember exactly the ambiguities, but Go allows if/for/switch/select to omit the parenthesis around
the conditional/iterator/... as in

  if true { ... }

However, this forces the grammar to introduce a special token
for the left brace of the then body (called loop_body in the grammar).

From golang spec there is also an issue with composite literal:

"A parsing ambiguity arises when a composite literal using the TypeName
form of the LiteralType appears as an operand between the keyword and
the opening brace of the block of an "if", "for", or "switch"
statement, and the composite literal is not enclosed in parentheses,
square brackets, or curly braces. In this rare case, the opening brace
of the literal is erroneously parsed as the one introducing the block
of statements. To resolve the ambiguity, the composite literal must
appear within parentheses."

if x == (T{a,b,c}[i]) { ... }
if (x == T{a,b,c}[i]) { ... }


In Go you have weird stuff like:
if func(...) bool { return ... }(...) { ... }
for a := struct {...} { ... } { ... }
for a := range []int{...} { ... }

**Comment by amaanq** [comment]
Could you rebase?

**Comment by josharian** [comment]
I know this is Alan's PR, but in case it is helpful, I rebased this commit and regenerated with `tree-sitter 0.20.8`. (Please do double-check my work if you want to use it.)

https://github.com/josharian/tree-sitter-go/commit/309e76588c33af4b9b48ea2b37751be01b21e253

Note that I found that regenerating at master with 0.20.8 made changes as well, so those would be mixed in here.

**Comment by amaanq** [comment]
If you want you can PR that, but remove the comments added, especially the LL one - but otherwise I think it looks good

**Comment by josharian** [comment]
Sent https://github.com/tree-sitter/tree-sitter-go/pull/124.

**Comment by amaanq** [comment]
Superseded by #124

**Pull request closed without merging**